File: python/exer02b04_producer_consumer.py
'''
THE PRODUCER-CONSUMER PROBLEM

SOLUTION TYPE B: USING SEMAPHORES
    Version B04: Multiple fast producers, multiple slow consumers
'''

import time
import threading



def producer(
    sem_fill: threading.Semaphore,
    sem_empty: threading.Semaphore,
    q: list,
    start_value: int
):
    time.sleep(1)
    i = 1

    while True:
        sem_empty.acquire()
        q.append(i + start_value)
        sem_fill.release()
        i += 1



def consumer(
    sem_fill: threading.Semaphore,
    sem_empty: threading.Semaphore,
    q: list
):
    while True:
        sem_fill.acquire()
        data = q.pop(0)
        print('Consumer', data)
        time.sleep(1)
        sem_empty.release()



s_fill = threading.Semaphore(0)  # item produced
s_empty = threading.Semaphore(1) # remaining space in queue
que = []

NUM_PRODUCERS = 3
NUM_CONSUMERS = 2

for i in range(NUM_PRODUCERS):
    threading.Thread(target=producer, args=(s_fill, s_empty, que, i * 1000)).start()

for _ in range(NUM_CONSUMERS):
    threading.Thread(target=consumer, args=(s_fill, s_empty, que)).start()

File: csharp/exercise/exer02b04-producer-consumer.cs
/*
 * THE PRODUCER-CONSUMER PROBLEM
 *
 * SOLUTION TYPE B: USING SEMAPHORES
 *      Version B04: Multiple fast producers, multiple slow consumers
 */
using System;
using System.Collections.Generic;
using System.Threading;



class Exer02B04 : IRunnable
{
    public void run()
    {
        var semFill = new Semaphore(0, int.MaxValue);     // item produced
        var semEmpty = new Semaphore(1, int.MaxValue);    // remaining space in queue

        var queue = new Queue<int>();

        const int NUM_PRODUCERS = 3;
        const int NUM_CONSUMERS = 2;


        var lstThProducer = new List<Thread>();
        var lstThConsumer = new List<Thread>();

        for (int i = 0; i < NUM_PRODUCERS; ++i)
        {
            int t = i;
            lstThProducer.Add(new Thread(() => producer(semFill, semEmpty, queue, t * 1000)));
        }

        for (int i = 0; i < NUM_CONSUMERS; ++i)
        {
            lstThConsumer.Add(new Thread(() => consumer(semFill, semEmpty, queue)));
        }


        lstThProducer.ForEach(th => th.Start());
        lstThConsumer.ForEach(th => th.Start());
    }


    private void producer(Semaphore semFill, Semaphore semEmpty,
        Queue<int> queue, int startValue)
    {
        int i = 1;

        for (; ; ++i)
        {
            semEmpty.WaitOne();
            queue.Enqueue(i + startValue);
            semFill.Release();
        }
    }


    private void consumer(Semaphore semFill, Semaphore semEmpty, Queue<int> queue)
    {
        for (; ; )
        {
            semFill.WaitOne();

            int data = queue.Dequeue();
            Console.WriteLine("Consumer " + data);
            Thread.Sleep(1000);

            semEmpty.Release();
        }
    }
}
